### PR TITLE
Support git dependencies

### LINKED ISSUE #1: Tracking issue for supporting git dependencies

### PR DESCRIPTION
**Pull request by lheckemann**
This adds support for git dependencies, which allows the generation of working expressions for more javascript projects.

Fixes #1

### CODE REVIEW

**Comment by zimbatm** [comment]
it looks like the nodejs in nixpkgs doesn't understand the `async` keyword yet. Is it possible to use promises instead?

File: bin/yarn2nix.js [hunk 1]
```diff
@@ -23,46 +24,70 @@ Options:
...
-  let found = {};
-
-  console.log(HEAD)
-
-  for (var depRange in lockedDependencies) {
-    let dep = lockedDependencies[depRange];
+async function fetchgit(url, sha1) {
+  let file_name = path.basename(url);
+  let url_for_git = url.replace(/^git\+/, "")
+  const {error, stdout, stderr} = await execFile("nix-prefetch-git", [url_for_git, sha1], {});
```
Comment (zimbatm) [inline bin/yarn2nix.js:37]: this means that using IDF + nix sandboxing won't work if git dependencies are introduced right?

**Comment by lheckemann** [comment]
>it looks like the nodejs in nixpkgs doesn't understand the async keyword yet. Is it possible to use promises instead?

That's odd, I didn't encounter this…
EDIT: It looks like this is the case in 18.03, but not 18.09 or unstable.

File: bin/yarn2nix.js (hunk 1 above)
Comment (lheckemann) [reply bin/yarn2nix.js:37]: Good point! Maybe having three modes — disallow git deps, use `nix-prefetch-git` (for "offline" generation, as we have here) and use `builtins.fetchGit` (for use with IFD) would make sense. `builtins.fetchGit` would also have the advantage of generating quickly, like when no git dependencies are involved.

**Comment by zimbatm** [comment]
Oops, I assumed the others were failing in the same way.

> error Can't make a request in offline mode ("http://registry.npmjs.org/minimist/-/minimist-1.2.0.tgz")

**Comment by lheckemann** [comment]
Is this using the mkYarnPackage from nixpkgs? It's not compatible with the master branch of yarn2nix.

**Comment by zimbatm** [comment]
I don't think it is?

**Comment by zimbatm** [comment]
I don't mean to hold this PR back, if you manage to make the CI go green I will merge happily

File: bin/yarn2nix.js (hunk 1 above)
Comment (srghma) [reply bin/yarn2nix.js]: `builtins.fetchGit` is must-need - it allows to download private repositories

**Comment by srghma** [comment]
I don't understand - which pr is more preferred? this or https://github.com/moretea/yarn2nix/pull/23 ?

File: default.nix [hunk 2]
```diff
@@ -292,7 +292,7 @@ in rec {
         rm -f .yarnrc
         cd $out/libexec/${pname}/deps/${pname}
         mkdir -p $out/tarballs/
-        yarn pack --ignore-scripts --filename $out/tarballs/${baseName}.tgz
+        yarn pack --offline --ignore-scripts --filename $out/tarballs/${baseName}.tgz
```
Comment (srghma) [inline default.nix]:
```
error Can't make a request in offline mode ("http://registry.npmjs.org/minimist/-/minimist-1.2.0.tgz")

```

there is an error during

File: bin/yarn2nix.js (hunk 1 above)
Comment (srghma) [inline bin/yarn2nix.js]: you can't call `nix-prefetch-git` during build phase https://github.com/NixOS/nixpkgs/issues/43093

I've tried and got an error 

```
(node:27308) UnhandledPromiseRejectionWarning: Error: Command failed: nix-prefetch-git https://github.com/srghma/node-shell-quote.git 0aa381896e0cd7409ead15fd444f225807a61e0a
Initialized empty Git repository in /tmp/nix-build-yarn.nix.drv-0/git-checkout-tmp-22QqXk01/node-shell-quote-0aa3818/.git/
From https://github.com/srghma/node-shell-quote
 * branch            newbranch  -> FETCH_HEAD
 * [new branch]      newbranch  -> origin/newbranch
Switched to a new branch 'fetchgit'
removing `.git'...
error: cannot open connection to remote store 'daemon': reading from file: Connection reset by peer
````

**Pull request closed without merging**